kymatio/utils.py

```diff
@@ -0,0 +1,59 @@
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class ScatteringTransformer(BaseEstimator, TransformerMixin):
+    def __init__(self, S, signal_shape, frontend='numpy'):
+        """Creates an object that is compatible with the scikit-learn API
+        and implements the `.transform` method.
+
+        Parameters
+        ==========
+
+        S: an instance of Scattering1D, Scattering2D or Scattering3D
+            This instance is called by the transformer.
+
+        signal_shape: tuple of ints
+            The shape of one sample. The `scikit-learn` convention is to work
+            with 2D arrays only of shape `(n_samples, n_features)`. Data is
+            delivered in this way and has to be reshaped before transforming.
+
+        frontend: string
+            The frontend of the `Scattering` object `S`.
+
+        Output
+        ======
+        Y: ndarray of shape (n_samples, n_scattering_features)
+            The scattering coefficients, raveled in row-major order (C-like).
+        """
+
+        self.scattering = S
+        self.signal_shape = signal_shape
+        self.frontend = frontend
+
+        assert frontend in ['numpy','torch']
+
+    def fit(self, X=None, y=None):
+        # no fitting necessary
+        return self
+
+    def predict(self, X):
+        if self.frontend == 'torch':
+            import torch
+
+        X_reshaped = X.reshape((-1,) + self.signal_shape)
+
+        if self.frontend is 'torch':
+            X_reshaped = torch.from_numpy(X_reshaped)
```

Comment:
if we're "frontend" independent, then, one should transform into np array any "type" of data? Maybe it's safer to be purely numpy compatible first? @janden

Reply:
Yes I don't see the why we should have this frontend stuff in here. As far as I can tell, `sklearn` is used with NP. Sure, there are extensions like `skorch` that let you do `sklearn` on `torch`, but this is a separate package.

Reply:
@edouardoyallon  the sklearn user never interacts with torch here except as a dependency, the transformer is np in np out. The default is also numpy so the torch and gpu acceleration are completely optional and only if the user really wants to use them.

If you can name me some real examples that arent just demos where scattering is useful on cpu esp in the context of sklearn it might be helpful to understand what we want from this.

Reply:
> gpu acceleration are completely optional and only if the user really wants to use them.

Yeah but how much acceleration do you get if you're moving it onto the GPU, transforming, and then off? Seems like you would spend a lot of time copying to and from the GPU.

> If you can name me some real examples that arent just demos where scattering is useful on cpu esp in the context of sklearn it might be helpful to understand what we want from this.

So all the audio experiments (classification, reconstruction, etc.) were done on CPU in a reasonable amount of time. For classification, it was hooked up to an SVM, which I think could nicely integrate into `sklearn`. Once this is merged, I could code up an example or two on some audio classification problems.

Reply:
> > gpu acceleration are completely optional and only if the user really wants to use them.
> 
> Yeah but how much acceleration do you get if you're moving it onto the GPU, transforming, and then off? Seems like you would spend a lot of time copying to and from the GPU.

In cases of reasonable sized natural images (e.g. 224x224x3) and larger J it would still give a huge speed up, 1 or 2 orders of magnitude even with data transfer. But ok its true 1d version might be reasonably fast for useful input sizes. 
> 
> > If you can name me some real examples that arent just demos where scattering is useful on cpu esp in the context of sklearn it might be helpful to understand what we want from this.
> 
> So all the audio experiments (classification, reconstruction, etc.) were done on CPU in a reasonable amount of time. For classification, it was hooked up to an SVM, which I think could nicely integrate into `sklearn`. Once this is merged, I could code up an example or two on some audio classification problems.

I'm  still have a worry that we should take care to not expose the casual user (e.g. someone trying to stick it in their sklearn pipeline) to something much slower than what they are used to, or what they get with another transformer. Perhaps we can print warnings or some such when the sizes are relatively large.

Reply:
> larger J it would still give a huge speed up, 1 or 2 orders of magnitude even with data transfer.

Ok that's good to know. In that case, it's something that would be useful more generally for the NP frontend. How about implementing torch (or TF) *backend* for NP that would enable these speedups.

> Perhaps we can print warnings or some such when the sizes are relatively large.

Eh I'm not too keen on this. We can put something in the docs on the NP frontend, but I don't think it's a good idea to warn for large sizes. It's not like calling the NP FFT with large inputs gives you a warning. If you work with large arrays, things are going to take longer (whether you're on GPU or CPU).

Reply:
Made an issue about GPU backends for NP. We can discuss adding warnings for large sizes in the issues.